exercises/complex-numbers/complex_numbers_test.cpp

```diff
@@ -0,0 +1,218 @@
+    const Complex c1{1.0, 2.0};
+    const Complex c2{3.0, 4.0};
+
+    REQUIRE(Complex(-5.0, 10.0) == c1 * c2);
+}
+
+TEST_CASE("imaginary_unit") {
+    const Complex c1{0.0, 1.0};
+    const Complex c2{0.0, 1.0};
+
+    REQUIRE(Complex(-1.0, 0.0) == c1 * c2);
+}
+
+TEST_CASE("divide_purely_real_numbers") {
+    const Complex c1{1.0, 0.0};
+    const Complex c2{2.0, 0.0};
+
+    REQUIRE(Complex(0.5, 0.0) == c1 / c2);
+}
+
+TEST_CASE("divide_purely_imaginary_numbers") {
+    const Complex c1{0.0, 1.0};
+    const Complex c2{0.0, 2.0};
+
+    REQUIRE(Complex(0.5, 0.0) == c1 / c2);
+}
+
+TEST_CASE("divide_numbers_with_real_and_imaginary_part") {
+    const Complex c1{1.0, 2.0};
+    const Complex c2{3.0, 4.0};
+
+    REQUIRE(Complex(0.44, 0.08) == c1 / c2);
+}
+
+TEST_CASE("absolute_value_of_a_positive_purely_real_number") {
+    const Complex c{5.0, 0.0};
+
+    REQUIRE(Approx(5.0) == c.abs());
+}
+
+TEST_CASE("absolute_value_of_a_negative_purely_real_number") {
+    const Complex c{-5.0, 0.0};
+
+    REQUIRE(Approx(5.0) == c.abs());
+}
+
+TEST_CASE(
+    "absolute_value_of_a_purely_imaginary_number_with_positive_imaginary_"
+    "part") {
+    const Complex c{0.0, 5.0};
+
+    REQUIRE(Approx(5.0) == c.abs());
+}
+
+TEST_CASE(
+    "absolute_value_of_a_purely_imaginary_number_with_negative_imaginary_"
+    "part") {
+    const Complex c{0.0, -5.0};
+
+    REQUIRE(Approx(5.0) == c.abs());
+}
+
+TEST_CASE("absolute_value_of_a_number_with_real_and_imaginary_part") {
+    const Complex c{3.0, 4.0};
+
+    REQUIRE(Approx(5.0) == c.abs());
+}
+
+TEST_CASE("conjugate_a_purely_real_number") {
+    const Complex c{5.0, 0.0};
+
+    REQUIRE(Complex(5.0, 0.0) == c.conj());
+}
+
+TEST_CASE("conjugate_a_purely_imaginary_number") {
+    const Complex c{0.0, 5.0};
+
+    REQUIRE(Complex(0.0, -5.0) == c.conj());
+}
+
+TEST_CASE("conjugate_a_number_with_real_and_imaginary_part") {
+    const Complex c{1.0, 1.0};
+
+    REQUIRE(Complex(1.0, -1.0) == c.conj());
+}
+
+TEST_CASE("real_part_of_a_purely_real_number") {
+    const Complex c{1.0, 0.0};
+
+    REQUIRE(Approx(1.0) == c.real());
+}
+
+TEST_CASE("real_part_of_a_purely_imaginary_number") {
+    const Complex c{0.0, 1.0};
+
+    REQUIRE(Approx(0.0) == c.real());
+}
+
+TEST_CASE("real_part_of_a_number_with_real_and_imaginary_number") {
+    const Complex c{1.0, 2.0};
+
+    REQUIRE(Approx(1.0) == c.real());
+}
+
+TEST_CASE("imaginary_part_of_a_purely_real_number") {
+    const Complex c{1.0, 0.0};
+
+    REQUIRE(Approx(0.0) == c.imag());
+}
+
+TEST_CASE("imaginary_part_of_a_purely_imaginary_number") {
+    const Complex c{0.0, 1.0};
+
+    REQUIRE(Approx(1.0) == c.imag());
+}
+
+TEST_CASE("imaginary_part_of_a_number_with_real_and_imaginary_part") {
+    const Complex c{1.0, 2.0};
+
+    REQUIRE(Approx(2.0) == c.imag());
+}
+
+TEST_CASE("eulers_identity") {
+    const Complex c{0.0, M_PI};
+    const std::complex<double> cx{0.0, M_PI};
+
+    REQUIRE(Complex(exp(cx).real(), exp(cx).imag()) == c.exp());
```

Comment:
This follows the pattern of other tests but is a bit dense.
It is clearer to read if we create an intermediate object? Something like this:

```c++
const std::complex<double> cx{1.0, 0.0};
const std::complex<double> expected {exp(cx)};
REQUIRE(Complex {expected.real(), expected.imag()} == c.exp());
```

Reply:
This is significantly different after changes. I implemented a constructor and `operator==` for interop with `std::complex<double>`. So the test now just reads:

``` cpp
const Complex c{0.0, M_PI};
const std::complex<double> cx{0.0, M_PI};

REQUIRE(exp(cx) == c.exp());
```

This is potentially too far though, since for this test they'll have to implement not only `exp()` but also at a minimum `bool operator==(const std::complex<double>&, const Complex&)`

Reply:
Hmmm, that is nicer to read.

I wonder if we could replace entirely replace `std::complex` with `Complex`? Several other languages seem to do that, such as [Java](https://github.com/exercism/java/blob/d59537b60e1a2b7243d0043d67255ecfcbbe2ffd/exercises/complex-numbers/src/test/java/ComplexNumberTest.java#L250), [Ruby](https://github.com/exercism/ruby/blob/3e9a042c07d4297e4bc6b243e6782221aa9298b0/exercises/complex-numbers/complex_numbers_test.rb#L174), [Python](https://github.com/exercism/python/blob/c05518984eba3f9ecdc88641545ba439f5117277/exercises/complex-numbers/complex_numbers_test.py#L169), and [JavaScript](https://github.com/exercism/javascript/blob/d4cc5f5aed4dbaf5bfd80d79ad7c29f226a39ccb/exercises/complex-numbers/complex-numbers.spec.js#L201).

If we do that, then we could remove two of the `==` operators and the `std::complex` constructor.

Another idea is to dispense with `==` entirely and compare the real and imaginary parts separately, like the Python and JavaScript do. At this point in the track, do you think using `std::numeric_limits` is too complex for students or an appropriate challenge?

Reply:
Oh right, we don't actually need `std::complex` at all. I like keeping the `==` operator at least, but yea I'll simplify so we can avoid `std::complex`. We'll still have all the operator implementations to do for this problem.

Using epsilon and scaling to the size of the operands isn't really necessary for most solutions. I expect most students will just use `std::abs(x-y) < .005` or similar if they need to do it, and the mentors can go as far down that rabbit hole as they want.

Reply:
Much improved! Looks good.

I like the idea that the students can submit a simpler solution using `std::abs` while the example shows a more detailed implementation.